ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchConfigurationHelper.java

```diff
@@ -229,40 +233,28 @@ private static String computeOSGiBundles(String bundleList, Map<String, List<IPl
 		if (bundles.get(IPDEBuildConstants.BUNDLE_SIMPLE_CONFIGURATOR) != null)
 			return "org.eclipse.equinox.simpleconfigurator@1:start"; //$NON-NLS-1$
 
-		StringBuilder buffer = new StringBuilder();
-		Set<String> initialBundleSet = new HashSet<>();
-		StringTokenizer tokenizer = new StringTokenizer(bundleList, ","); //$NON-NLS-1$
+		Map<IPluginModelBase, String> allBundles = new HashMap<>(bundlesWithStartLevels);
+		StringTokenizer tokenizer = new StringTokenizer(initialBundleList, ","); //$NON-NLS-1$
 		while (tokenizer.hasMoreTokens()) {
 			String token = tokenizer.nextToken();
 			int index = token.indexOf('@');
 			String id = index != -1 ? token.substring(0, index) : token;
-			if (bundles.containsKey(id)) {
-				if (buffer.length() > 0)
-					buffer.append(',');
-				buffer.append(id);
-				if (index != -1 && index < token.length() - 1)
-					buffer.append(token.substring(index));
-				initialBundleSet.add(id);
+			IPluginModelBase model = getLatestModel(id, bundles);
+			if (model != null) {
+				String startInfo = index + 1 < token.length() ? token.substring(index + 1) : ""; //$NON-NLS-1$
+				allBundles.put(model, startInfo);
 			}
 		}
-
-		// write out all bundles in osgi.bundles - bug 170772
-		initialBundleSet.add(IPDEBuildConstants.BUNDLE_OSGI);
 		for (Entry<IPluginModelBase, String> entry : bundlesWithStartLevels.entrySet()) {
 			IPluginModelBase model = entry.getKey();
 			String id = model.getPluginBase().getId();
-			if (!initialBundleSet.contains(id)) {
-				if (buffer.length() > 0)
-					buffer.append(',');
-
-				String slinfo = entry.getValue();
-				buffer.append(id);
-				buffer.append('@');
-				buffer.append(slinfo);
+			if (!allBundles.containsKey(model) && //
+					!IPDEBuildConstants.BUNDLE_OSGI.equals(id)) { // write out all bundles in osgi.bundles - bug 170772
+				allBundles.put(model, entry.getValue());
 			}
 		}
 
-		return buffer.toString();
+		return allBundles.entrySet().stream().map(entry -> getBundleURL(entry.getKey(), true) + getStartData(entry.getValue(), defaultAuto)).collect(Collectors.joining(",")); //$NON-NLS-1$
```

Comment:
For better readability I would instead of this one-liner would rather go with
```suggestion
		return allBundles.entrySet().stream().map(entry -> {
			IPluginModelBase bundle = entry.getKey();
			String startInfo = entry.getValue();
			return getBundleURL(bundle, true) + getStartData(startInfo, defaultAuto);
		}).collect(Collectors.joining(",")); //$NON-NLS-1$
```
or
```suggestion
		StringJoiner osgiBundles = new StringJoiner(","); //$NON-NLS-1$
		allBundles.forEach((bundle, startInfo) -> {
			String bundleURL = getBundleURL(bundle, true);
			String startData = getStartData(startInfo, defaultAuto);
			osgiBundles.add(bundleURL + startData);
		});
```